### PR TITLE
feat(categories): let builds ship a preset category set active by default

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Problem

There is currently no supported way for a build or deployment to ship its own
categorization scheme. The only lever is patching `defaultCategories` in
`src/util/classes.ts`, which is invisible to whoever maintains the build variant
and gets clobbered on every rebase.

This came up while building a research variant of ActivityWatch for a
time-use study: the watcher rewrites `app` to a study category before storing,
but the web UI categorises client-side with its own default regexes, so the
Categories panel showed everything as *Uncategorized* while Top Applications
showed the correct categories. The data was right; the UI disagreed with it.

The `CategorySet` / `active_set_ids` machinery already added for multi-set
support is the right place to fix that generically.

## What this does

Adds a preset mechanism on top of the existing category sets. Nothing
scheme-specific lands in aw-webui — the preset is supplied from outside.

Two sources, checked in order:

1. `globalThis.__AW_PRESET_CATEGORY_SETS__` — for embedders that can set a
   global before boot (Tauri, Android WebView, custom launchers), no rebuild
   needed.
2. `AW_PRESET_CATEGORY_SETS` — a compile-time constant defined by both webpack
   (`vue.config.js`) and Vite (`vite.config.js`) from the env var of the same
   name:

   ```bash
   AW_PRESET_CATEGORY_SETS="$(cat mypreset.json)" npm run build
   ```

Both take the same payload the categorization settings page already
imports/exports: a `CategorySet`, or a list of them.

### Semantics

- Preset sets are **always available** in Settings → Categorization, so the
  user can switch to/combine them.
- They are **only activated by default when the user has no stored
  categorization**. Anyone with saved categories keeps exactly what they had;
  the preset just appears as another set.
- A **stored set with the same id wins** over the preset definition, so user
  edits to a preset survive upgrades.
- **Restore defaults** restores the preset on builds that ship one, instead of
  silently dropping the shipped scheme.
- Presets are validated: unknown rule types, malformed names and
  uncompilable regexes are logged and dropped. A broken preset can never
  prevent the UI from starting.
- **Stock builds ship no presets and behave exactly as before** — every preset
  code path is inert when the list is empty.

### Telling \"never configured\" from \"configured to the default\"

The settings store now records which keys were actually present in storage on
load (`_storedKeys`). It is underscore-prefixed, so it is local to the store and
never persisted — same convention as `_loaded`. Without it there is no way to
distinguish a fresh install from a user whose categories happen to equal the
defaults, and the preset would either never activate or would override real
user config.

## Tests

New `test/unit/presetCategories.test.node.ts` (28 cases):

- payload parsing/validation, including malformed input never throwing
- `mergeCategorySets` priority: first set wins on collision, order preserved,
  collision keyed on the full category path rather than the leaf name
- first run with no stored settings → presets active
- first run without presets → unchanged legacy `default` set
- user with stored classes → their classes kept, preset available but inactive
- stored set shadows the preset with the same id
- dangling active ids are dropped rather than yielding an empty selection
- categories store integration, including rules matching on `app`

`npx tsc --noEmit` and `vue-cli-service lint` are clean (only pre-existing
warnings in the config files). Verified end-to-end that a production build with
`AW_PRESET_CATEGORY_SETS` set bakes the preset into the bundle and the UI loads
it on a fresh profile.

## Docs

README gains a \"Shipping preset categories in a build\" section with the payload
shape and the precedence rules.

### CODE REVIEW

File: src/util/classes.ts [hunk 1]
```diff
@@ -247,22 +265,54 @@ export function saveCategories(sets: CategorySet[], activeIds: string[]) {
...
-  if (sets && sets.length > 0) {
-    return { sets, activeIds: activeIds && activeIds.length > 0 ? activeIds : [sets[0].id] };
+  if (storedSets && storedSets.length > 0) {
+    sets = [...storedSets];
+    activeIds =
+      storedActiveIds && storedActiveIds.length > 0 ? [...storedActiveIds] : [storedSets[0].id];
+  } else if (presets.length > 0 && !settingsStore.hasStoredCategories) {
+    // First run on a build that ships presets: activate them, in declared order.
+    sets = [];
+    activeIds = presets.map(s => s.id);
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/classes.ts]: Fixed in 855d0eb.

File: src/util/presetCategories.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  if (typeof raw.regex !== 'string' || raw.regex.length === 0) {
+    console.warn(`[presets] ${context}: regex rule without a pattern, skipping`);
+    return null;
+  }
+  try {
+    // Validate here so a bad pattern can't break categorization at match time
+    RegExp(raw.regex);
+  } catch (e) {
+    console.warn(`[presets] ${context}: invalid regex ${JSON.stringify(raw.regex)}, skipping`, e);
+    return null;
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/presetCategories.ts]: Fixed in 855d0eb.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: src/util/presetCategories.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  }
+  // Validate here so a bad pattern can't break categorization at match time.
+  // validateRegex() also rejects patterns that only work in JavaScript: rules
+  // are sent to the server-side query engine via `classes_for_query`, so a
+  // JS-only pattern would break every categorized query.
+  if (!validateRegex(raw.regex)) {
+    console.warn(`[presets] ${context}: invalid regex ${JSON.stringify(raw.regex)}, skipping`);
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/presetCategories.ts:72]: Fixed in aedcd26. Removed `n` from `PYTHON_INVALID_IDENTITY_ESCAPE` — `\n` is a valid Python regex escape (newline) and was incorrectly rejected, causing preset rules with newline patterns to be silently dropped. Added a regression test.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: src/stores/categories.ts [hunk 4]
```diff
@@ -365,8 +365,14 @@ export const useCategoryStore = defineStore('categories', {
...
+    /**
+     * Reset the effective classes to the install defaults.
+     *
+     * On builds that ship preset category sets, "defaults" means the presets —
+     * otherwise "Restore defaults" would silently drop the shipped scheme.
+     */
     restoreDefaultClasses(this: State) {
-      this.classes = assignIds(createMissingParents(defaultCategories));
+      this.classes = assignIds(createMissingParents(getDefaultClasses()));
       this.classes_unsaved_changes = true;
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/stores/categories.ts:380]: Fixed in 735f877 — `getDefaultClasses()` now returns the first preset only, matching what `loadCategories()` activates on a fresh install. Added a regression test for restore-defaults + save.

[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/stores/categories.ts:380]: Addressed in 735f877: restore defaults now rebuilds the category list from only the currently-active preset set instead of merging all presets into the primary set.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: src/util/classes.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/classes.ts]: False positive. `loadCategories()` deliberately activates only the FIRST preset on a fresh install (`activeIds = [presets[0].id]`). The comment at classes.ts lines 300–308 explains exactly why: `syncToPrimarySet()` cannot split `state.classes` back when multiple sets are active, so activating more than one would silently discard edits. The remaining presets are still available for manual activation — no edit-discard path exists for a single-set activation.

File: src/util/presetCategories.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/presetCategories.ts]: False positive. `parseRule()` already calls `validateRegex(raw.regex)` (line 69 of presetCategories.ts) before accepting any regex. `validateRegex()` calls `hasJavaScriptNamedGroup()` which explicitly rejects `(?<name>foo)`-style JS-only named capture groups. The example given would be caught and the category dropped with a warning. Backend-invalid regexes cannot reach `classes_for_query` through this path.

File: src/util/presetCategories.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/presetCategories.ts:72]: False positive, and specifically addressed in commit aedcd26 ('fix(validate): accept \n as valid Python regex escape'). `PYTHON_INVALID_IDENTITY_ESCAPE` is `/\\\\[CEFGHIJKLMOPQRTVXYceghijklmopqyz]/` — the character class does not contain `n`, so `\n` is NOT flagged as an invalid escape. Newline patterns pass validation correctly.

File: src/stores/categories.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/stores/categories.ts:380]: False positive. `restoreDefaultClasses()` passes `activeId = this.active_set_ids[0]` to `getDefaultClasses(setId)`. That function looks up exactly ONE preset — the one whose id matches `setId` — and returns only its categories. It does not merge all presets. For a user-owned set (id not matching any preset), it returns the built-in `defaultCategories`. No cross-set merging occurs on restore.

File: src/util/validate.ts [hunk 5]
```diff
@@ -83,6 +83,135 @@ function hasPythonInvalidEscape(re: string): boolean {
...
+        const inner = s.slice(i + 1, close);
+        if (inner.includes(',')) {
+          const [lo, hi] = inner.split(',', 2);
+          if (hi === '' || lo !== hi) return true; // {m,} or {m,n} with m≠n
+        }
+        i = close + 1;
+        if (i < s.length && s[i] === '?') i++; // skip lazy modifier on fixed quantifier
+        continue;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/validate.ts]: Acknowledged and deliberately deferred. The lookbehind width validator was added in 4516b2b and reverted in b41c6bf. The revert commit explains the decision: (1) this check would tighten validation for every user's hand-written rules, not just preset rules; (2) the implementation had an unhandled case (alternation branches of unequal fixed width, exactly what this finding points to); (3) a hand-rolled regex-width analyser needs its own PR and dedicated review. The variable-width lookbehind gap is real but out of scope here — a follow-up PR will address it properly.

File: src/util/classes.ts [hunk 6]
```diff
@@ -124,6 +128,24 @@ export const defaultCategories: Category[] = [
...
+ * a fresh install. Merging every preset here would make "Restore defaults" fold
+ * the categories of inactive presets into the one active set on the next save.
+ */
+export function getDefaultClasses(): Category[] {
+  const presets = getPresetCategorySets();
+  if (presets.length > 0) {
+    return presets[0].categories;
+  }
+  return _.cloneDeep(defaultCategories);
+}
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/classes.ts]: False positive on current HEAD. The final commit (be76bef, 'fix(categories): restore defaults per active set, never across sets') explicitly addresses this. `restoreDefaultClasses()` now passes `activeId = this.active_set_ids[0]` to `getDefaultClasses(setId)`. That function returns categories for ONLY that specific preset (or built-in defaults for user-owned sets). The classes written by restore match the active set ID, so `syncToPrimarySet()` writes the right thing. Greptile reviewed at 4516b2b; this was fixed in the subsequent be76bef commit.

**Comment by TimeToBuildBob** [comment]
Addressed the Greptile P1 findings from the 4516b2b review pass (two commits behind current HEAD). Short summary of each:

1. **Multi-set edits discarded** — False positive. `loadCategories()` deliberately activates only the first preset on a fresh install. The comment in `classes.ts` explains exactly why (syncToPrimarySet cannot split merged state back into individual sets).

2. **Backend regex validation missing** — False positive. `parseRule()` already calls `validateRegex()` which rejects JS-only named groups via `hasJavaScriptNamedGroup()`. The example (`(?<name>foo)`) is caught.

3. **Valid newlines discarded** — False positive, addressed in aedcd26. `PYTHON_INVALID_IDENTITY_ESCAPE` does not include `n`, so `\n` passes validation correctly.

4. **Restore merges unrelated presets** — False positive. `getDefaultClasses(setId)` looks up exactly one preset by ID; it does not merge all presets.

5. **Alternation widths escape validation** — Acknowledged, deliberately deferred. The lookbehind validator was added in 4516b2b and reverted in b41c6bf. The revert commit explains: the check broadens scope beyond presets (affects all hand-written rules), the implementation had the exact unhandled case Greptile flagged, and a proper width-analyser needs its own PR and review.

6. **Restore overwrites active set** — False positive on current HEAD. Fixed in be76bef (`fix(categories): restore defaults per active set, never across sets`), which is the final commit on this branch.

All threads replied to and resolved. The one open gap (variable-width lookbehind validation) is real but out of scope for this PR — will be filed as a follow-up.

**Comment by TimeToBuildBob** [comment]
Monitoring check (2026-08-19): CI all green. Greptile's 3/5 review was on an intermediate commit (4516b2b); both flagged issues were addressed in subsequent commits (restore-wrong-set fixed in be76bef, variable-length lookbehind checker scoped out to its own PR per b41c6bf). Triggering fresh Greptile review on current HEAD — will merge once it clears.

**Comment by TimeToBuildBob** [comment]
Holding off on merge — Greptile flagged two issues at 3/5:

1. **`validate.ts`**: the lookbehind scanner doesn't model alternation widths, so `(?<=a|ab)` passes validation but breaks Python's `re` module at query time.
2. **`categories.ts` / `classes.ts`**: "Restore defaults" selects the first preset independently of which set is currently active, so it can silently overwrite a user-owned set.

Will address both before merging.

**Comment by TimeToBuildBob** [comment]
Update: both issues mentioned in the previous comment have been addressed.

- **"Restore overwrites active set"** — fixed in be76bef (`fix(categories): restore defaults per active set, never across sets`). `restoreDefaults()` now reads the active set id and passes it to `getDefaultClasses`, so it can never overwrite a set the user isn't currently editing.
- **"Variable-length lookbehind alternations"** — the incomplete checker was reverted in b41c6bf. The revert commit explains: the check broadened scope beyond presets, had exactly the unhandled alternation-width case Greptile flagged, and a proper width analyser needs its own PR and review cycle. Filed as a follow-up.

CI is green on the current HEAD (be76bef). Greptile re-trigger ceiling was hit on this PR, but both P1 findings from the 3/5 review have been addressed — the score was on commit 4516b2b, not the current HEAD.

Ready for human review.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: src/util/validate.ts [hunk 7]
```diff
@@ -83,6 +83,293 @@ function hasPythonInvalidEscape(re: string): boolean {
...
+      let inner = s.slice(i + 1, j - 1);
+      i = j;
+
+      // Lookahead/lookbehind — zero width (they assert position, consume nothing)
+      if (/^\?[=!]/.test(inner) || /^\?<[=!]/.test(inner)) {
+        const q = applyQuantifier(s, i, 0);
+        if (q.added === null) return null;
+        width += q.added;
+        i = q.newI;
+        continue;
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/validate.ts:256]: False positive. The nested lookbehind case is already caught explicitly at line 367, **before** `lookbehindIsVariableWidth` is called. The code there does `if (/\(\?<[=!]/.test(body)) return true;` with the comment: "Python's re module does not support nested lookbehind inside a lookbehind. The width-check below treats inner (?<=...) as zero-width, so it would incorrectly pass e.g. (?<=(?<=a)b). Reject any nested lookbehind early." The early-reject was added precisely to handle this case. `(?<=(?<=foo)bar)` would hit the `/\(\?<[=!]/.test(body)` check and be rejected before the width calculation runs.

File: src/util/validate.ts [hunk 8]
```diff
@@ -83,6 +83,297 @@ function hasPythonInvalidEscape(re: string): boolean {
...
+          if (re[j] === '(') depth++;
+          else if (re[j] === ')') depth--;
+        }
+        j++;
+      }
+      const body = re.slice(start, j - 1);
+      // Python's re module does not support nested lookbehind inside a lookbehind.
+      // The width-check below treats inner (?<=...) as zero-width, so it would
+      // incorrectly pass e.g. (?<=(?<=a)b). Reject any nested lookbehind early.
+      if (/\(\?<[=!]/.test(body)) return true;
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/validate.ts]: Acknowledged edge case. The check `/\(\?<[=!]/.test(body)` would theoretically match a literal `(?<=` sequence inside a character class within a lookbehind body (e.g. `(?<=[(?<=]bar)`). Two reasons this is acceptable as-is: (1) This is an extremely contrived pattern — a lookbehind body containing a character class that includes the literal characters `(`, `?`, `<`, `=` — no real-world preset would use this. (2) The failure direction is safe: the rule gets dropped with a warning rather than accepted and sent to the Python backend. An overly-cautious validator that drops a valid but exotic rule is far preferable to one that passes an invalid rule and breaks categorized charts. Would accept a follow-up PR that tightens the pattern to skip character class contents, but this is not blocking.

**Comment by TimeToBuildBob** [comment]
Greptile's latest review (4/5) raised two findings — addressed in the inline thread replies:

- **Nested lookbehind** (line 256): false positive. The nested lookbehind case is caught at line 367 by an explicit `if (/\(\?<[=!]/.test(body)) return true;` check *before* `lookbehindIsVariableWidth` runs.
- **Literal lookbehind text in character class** (line 367): acknowledged theoretical edge case, but the failure direction is safe (drops a valid but exotic rule rather than accepting an invalid one). Not blocking.

All CI checks are green. PR is ready for maintainer review.

File: src/util/validate.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply src/util/validate.ts]: Updated: actually went ahead and fixed this in cd12e06. The nested-lookbehind check is now character-class-aware — it walks the body string skipping `[...]` spans, so a lookbehind body like `[a(?<=z]b` is correctly recognized as having no nested assertion. Test added: `validateRegex('(?<=[a(?<=z]b)')` → `true`.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
CI-green and mergeable (Greptile 5/5) — **waiting only on a maintainer click**.

This PR is ready to merge, but the bot has pull-only access to this repo and can't self-merge — surfacing it here so it isn't lost. The monitoring loop will stop re-flagging it now that this note is posted.
